### ISSUE TITLE
Debugger API documentation 

### ISSUE THREAD

**Issue by fatcerberus** [Microsoft/ChakraCore#3636]
Is there any centralized place where I can find documentation for the JSON-based debug API?  Some of the JSON is documented in ChakraDebug.h but not all--for example there's no documentation for the JSON object that gets passed to the event callback.

**Comment by liminzhu** [comment]
No at the moment, but I will get the debugging APIs in the API reference docs. For the time being, you can see the [ChakraDebug.h](https://github.com/Microsoft/ChakraCore/blob/master/lib/Jsrt/ChakraDebug.h) should have the info you need, including the returned JSON (see remarks).

**Comment by fatcerberus** [comment]
The only thing I couldn't find in `ChakraDebug.h` was the JSON passed to the event callback (`eventData` parameter).  Based on experimentation I figured out it has a scriptId along with line and column information, but beyond that I have no clue.  I haven't tried serializing it yet to see what it contains.

**Comment by agarwal-sandeep** [comment]
@fatcerberus below is the json which is returned for debugEvents. The data may vary based on scenario but the structure is same.
```javascript
{
  "debugEvent": "JsDiagDebugEventSourceCompile",
  "eventData": {
    "scriptType": "eval code",
    "lineCount": 1,
    "sourceLength": 10,
    "scriptId": 4,
    "parentScriptId": 2
  }
}, {
  "debugEvent": "JsDiagDebugEventBreakpoint",
  "eventData": {
    "breakpointId": 2,
    "scriptId": 2,
    "line": 2,
    "column": 2,
    "sourceLength": 9,
    "sourceText": "var y = 2"
  }
}, {
  "debugEvent": "JsDiagDebugEventStepComplete",
  "eventData": {
    "scriptId": 2,
    "line": 3,
    "column": 2,
    "sourceLength": 9,
    "sourceText": "var z = 1"
  }
}, {
  "debugEvent": "JsDiagDebugEventDebuggerStatement",
  "eventData": {
    "scriptId": 2,
    "line": 4,
    "column": 2,
    "sourceLength": 8,
    "sourceText": "debugger"
  }
}
```

**Comment by fatcerberus** [comment]
Thanks, and for `JsDiagDebugEventRuntimeException`?

**Comment by liminzhu** [comment]
Assigning myself to this so that I don' forget updating the docs :).

**Comment by agarwal-sandeep** [comment]
```javascript
{
  "debugEvent": "JsDiagDebugEventRuntimeException",
  "eventData": {
    "scriptId": 2,
    "line": 3,
    "column": 4,
    "sourceLength": 5,
    "sourceText": "bar()",
    "uncaught": false,
    "exception": {
      "name": "{exception}",
      "type": "object",
      "className": "Error",
      "display": "'bar' is not defined",
      "propertyAttributes": 3,
      "handle": 1
    }
  }
}
```

**Comment by fatcerberus** [comment]
@agarwal-sandeep Thanks.  Interestingly, for a runtime exception event, `exception.handle` *always* seems to be 1 -- which apparently doesn't refer to the actual exception value that was thrown.  I tried resolving the handle to an object and accessing `.stack` from it and it didn't work.  Is that a bug?

**Comment by agarwal-sandeep** [comment]
@fatcerberus handle is a global counter which starts from 1, if no JsDiag* APIs where called before JsDiagDebugEventRuntimeException got handled the handle referred in exception object will get handle 1. Based on code [here](https://github.com/Microsoft/ChakraCore/blob/5a9d0ee1bc25af874e11de5f8999d030ee2961de/lib/Jsrt/JsrtDebugManager.cpp#L334) it shouldn't be a bug but I will look into it more.

In JsDiag* APIs actual JavaScript object or properties are not exposed to host. The actual exception object is saved as internal object in the exception object ("name" : "exception") we generate to pass in eventData. You should be able to call [JsDiagGetProperties](https://github.com/Microsoft/ChakraCore/blob/f2078fed54fdeca4c3842fbd8d007de2502b06d5/lib/Jsrt/ChakraDebug.h#L521) to get all properties of the actual exception object ([obj](https://github.com/Microsoft/ChakraCore/blob/f2078fed54fdeca4c3842fbd8d007de2502b06d5/lib/Jsrt/JsrtDebugManager.cpp#L327))

**Comment by fatcerberus** [comment]
I used `JsDiagGetObjectFromHandle()` to get the object from the handle, and passing the handle from the event data gives me back at least `"name": "{exception}"` -- so it appears that the handle refers to the JSON exception object from the event data itself rather than the real exception that was thrown.

**Comment by agarwal-sandeep** [comment]
@fatcerberus Yes JsDiagGetObjectFromHandle will get you the exception object passed to eventData but JsDiagGetProperties knows how to get the children's from the  object as the [display](https://github.com/Microsoft/ChakraCore/blob/f2078fed54fdeca4c3842fbd8d007de2502b06d5/lib/Runtime/Debug/DiagObjectModel.cpp#L136) use internal exception object (obj) to get the [properties](https://github.com/Microsoft/ChakraCore/blob/f2078fed54fdeca4c3842fbd8d007de2502b06d5/lib/Jsrt/JsrtDebuggerObject.cpp#L575)

**Comment by fatcerberus** [comment]
Thanks, that's a bit roundabout but I know what to now.  I guess the debugger is implemented that way to avoid unwanted side effects from getters, proxies, etc.  Males sense. :)

**Comment by dilijev** [comment]
@liminzhu is this resolved?

**Comment by liminzhu** [comment]
not yet
